components/roster/heroes/Heroes.tsx

```diff
@@ -0,0 +1,283 @@
+"use client";
+
+import { FormEvent, useReducer } from "react";
+import { Hero } from "@/lib/roster";
+import { Card, CardContent } from "@/components/ui/card";
+import { AscensionLevel } from "@/lib/characters";
+import {
+  DropdownMenu,
+  DropdownMenuContent,
+  DropdownMenuItem,
+  DropdownMenuLabel,
+  DropdownMenuSeparator,
+  DropdownMenuSub,
+  DropdownMenuSubContent,
+  DropdownMenuSubTrigger,
+  DropdownMenuTrigger,
+} from "@/components/ui/dropdown-menu";
+import { Checkbox } from "@/components/ui/checkbox";
+import { Label } from "@/components/ui/label";
+import HeroPortrait from "@/components/roster/heroes/HeroPortrait";
+import { Check, Star, Unlock, Lock } from "lucide-react";
+import Image from "next/image";
+import { rosterImages } from "@/lib/images";
+import { Button } from "@/components/ui/button";
+import { toast } from "sonner";
+
+type SubSelectProps = {
+  hero: Hero;
+  heroProperty: "ascension" | "equipment";
+  options: AscensionLevel[] | string[];
+  selected: AscensionLevel | string | null | undefined;
+  label: string;
+  disabled: boolean;
+};
+
+type HeroProps = {
+  heroList: Hero[];
+};
+
+const isAscensionLevelValid = (level: AscensionLevel | null): boolean => {
+  if (!level) return false;
+  const validLevels = Object.keys(AscensionLevel).splice(7);
+  return validLevels.includes(level);
+};
+
+const exclusiveEquipmentOptions = Array.from(
+  { length: 6 },
+  (_, i) => `${i * 5}`,
+);
+
+type HeroChange = {
+  property: "unlock" | "ascension" | "equipment";
+  value: string;
+  hero: Hero;
+};
+
+function reducer(heroes: Hero[], change: HeroChange) {
+  const property = change.property;
+  const value = change.value;
+  const hero = change.hero;
+
+  switch (property) {
+    case "unlock": {
+      return heroes.map((h) =>
+        h.key === hero.key
+          ? { ...h, unlocked: JSON.parse(value.toLowerCase()) }
+          : h,
+      );
+    }
+    case "ascension": {
+      let ascensionLevel = (
+        Object.keys(AscensionLevel) as Array<keyof typeof AscensionLevel>
+      ).find((key) => AscensionLevel[key] === value) as AscensionLevel;
+
+      if (!isAscensionLevelValid(ascensionLevel)) {
+        return heroes.map((h) =>
+          h.key === hero.key
+            ? { ...h, ascension: ascensionLevel, exEquipment: 0 }
+            : h,
+        );
+      }
+
+      return heroes.map((h) =>
+        h.key === hero.key ? { ...h, ascension: ascensionLevel } : h,
+      );
+    }
+    case "equipment": {
+      return heroes.map((h) =>
+        h.key === hero.key ? { ...h, exEquipment: parseInt(value) } : h,
+      );
+    }
+    default:
+      return heroes;
+  }
+}
+
+export default function Heroes({ heroList }: HeroProps) {
+  const [heroes, dispatch] = useReducer(reducer, heroList);
```

Comment:
see comment about local state variables. i don't think using a reducer is necessary for this.

Reply:
reducer is fine here, ignore.